# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/BytecodeAnalysis.cpp
# Commit: f5e7bea40350
# Full Hash: f5e7bea40350979b1009753828e77b25e94b43e1
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-31 09:35:41
# Regressor Bug: 885514
# File Overlap Count: 1
# Description:
#   Bug 885514: Support finally in Warp r=jandem
#   
#   This supports any try-finally that doesn't emit a `JSOp::Retsub`. Currently that means that we don't support break, continue, or return inside the try block, but there's room to fix that in the future.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D142119
# ==============================================================================

diff -r 7f75f8b705e7 -r f5e7bea40350 js/src/jit/BytecodeAnalysis.cpp
--- a/js/src/jit/BytecodeAnalysis.cpp	Wed Mar 30 23:19:47 2022 +0000
+++ b/js/src/jit/BytecodeAnalysis.cpp	Wed Mar 30 23:19:48 2022 +0000
@@ -31,11 +31,14 @@
   mozilla::PodZero(infos_.begin(), infos_.length());
   infos_[0].init(/*stackDepth=*/0);
 
-  // Because WarpBuilder can compile try-blocks but doesn't compile the
-  // catch-body, we need some special machinery to prevent OSR into Warp code in
-  // the following cases:
+  // WarpBuilder can compile try blocks, but doesn't support handling
+  // exceptions. If exception unwinding would resume in a catch or finally
+  // block, we instead bail out to the baseline interpreter. Finally blocks can
+  // still be reached by normal means, but the catch block is unreachable and is
+  // not compiled. We therefore need some special machinery to prevent OSR into
+  // Warp code in the following cases:
   //
-  // (1) Loops in catch/finally blocks:
+  // (1) Loops in catch blocks:
   //
   //       try {
   //         ..
@@ -43,7 +46,7 @@
   //         while (..) {} // Can't OSR here.
   //       }
   //
-  // (2) Loops only reachable via a catch/finally block:
+  // (2) Loops only reachable via a catch block:
   //
   //       for (;;) {
   //         try {
@@ -55,8 +58,8 @@
   //       while (..) {} // Loop is only reachable via the catch-block.
   //
   // To deal with both of these cases, we track whether the current op is
-  // 'normally reachable' (reachable without going through a catch/finally
-  // block). Forward jumps propagate this flag to their jump targets (see
+  // 'normally reachable' (reachable without exception handling).
+  // Forward jumps propagate this flag to their jump targets (see
   // BytecodeInfo::jumpTargetNormallyReachable) and when the analysis reaches a
   // jump target it updates its normallyReachable flag based on the target's
   // flag.
@@ -281,10 +284,6 @@
         result.usesEnvironmentChain = true;
         break;
 
-      case JSOp::Finally:
-        result.hasTryFinally = true;
-        break;
-
       default:
         break;
     }